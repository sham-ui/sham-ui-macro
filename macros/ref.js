const t = require( 'babel-types' );

/**
 * Macro for get/generate uniq reference
 * @function ref
 * @param {string} [name] Name of reference
 *
 * @example
 * // Config in package.json
 * // "babelMacros": {
 * //   "ref": {
 * //     "enabled": true,
 * //     "uniq": true
 * //   }
 * // }
 *
 * import { ref } from 'sham-ui-macro/ref.macro';
 *
 * const firstName = ref();
 * const lastName = ref();
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ ref( 'firstName' ) ] + state[ ref( 'lastName' ) ];
 *
 * // ↓ ↓ ↓ ↓ ↓ ↓
 *
 * const firstName = 0;
 * const lastName = 1;
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ 0 ] + state[ 1 ];
 */

/**
 * @param {BabelPath} callExpressionPath
 * @param {Object} config
 * @param {Object} state
 * @private
 */
module.exports = function ref( callExpressionPath, config, state ) {
    const { uniq = false } = config;

    let name;
    const args = callExpressionPath.node.arguments;
    if ( 0 === args.length  ) {
        const parentPath = callExpressionPath.parentPath;
        if ( !parentPath.isVariableDeclarator() ) {
            throw new Error( 'Pass name to ref' );
        }

        // Use variable as name
        name = parentPath.node.id.name;
    } else {
        name = args[ 0 ].value;
    }

    let identifier;
    if ( uniq ) {
        if ( !state.map.has( name ) ) {
            state.map.set(
                name,
                state.map.size
            );
        }
        identifier = t.NumericLiteral( state.map.get( name ) );
    } else {
        identifier = t.StringLiteral( name );
    }

    callExpressionPath.replaceWith(
        identifier
    );
};
