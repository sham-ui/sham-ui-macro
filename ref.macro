/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const ref = require( './macros/ref' );

const refState = {
    map: new Map()
};

module.exports = createMacro( macro, {
    configName: 'ref'
} );

function macro( { references, config } ) {
    const { enabled = false } = config;
    if ( enabled ) {

        // After transpile with rollup in cjs import will:
        // var ref_macro = require( 'sham-ui/ref.macro' );
        // ...
        // const a = ref_macro.ref();
        // ...
        // But for ES modules transpile to
        // import { ref } from 'sham-ui/ref.macro';
        ( references.default || references.ref ).forEach( referencePath => {
            ref(
                referencePath.findParent( x => x.isCallExpression() ),
                config,
                refState
            )
        } );
    } else {

        // Keep imports for process in app level
        return {
            keepImports: true
        }
    }
}
