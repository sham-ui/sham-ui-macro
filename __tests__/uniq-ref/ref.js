/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: 'without args',
        code: `
            import { ref } from '../../ref.macro';

            const foo = ref();
            const bar = ref();
        `
    }, {
        title: 'with args',
        code: `
            import { ref } from '../../ref.macro';

            const foo = ref( 'foo' );
            const foo1 = ref( 'foo' );
        `
    }, {
        title: 'as key',
        code: `
            import { ref } from '../../ref.macro';
            const foo = ref();
            const bar = ref();

            const state = { [ ref( 'foo' ) ]: true, [ bar ]: false }; 
        `
    }, {
        title: 'don\'t used',
        code: `
            import '../../ref.macro';
        `
    } ]
} );
