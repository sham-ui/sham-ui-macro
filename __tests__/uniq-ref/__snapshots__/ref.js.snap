// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros as key: as key 1`] = `

import { ref } from '../../ref.macro';
const foo = ref();
const bar = ref();

const state = { [ ref( 'foo' ) ]: true, [ bar ]: false };

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = 0;
const bar = 1;

const state = { [0]: true, [bar]: false };

`;

exports[`macros don't used: don't used 1`] = `

import '../../ref.macro';

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`macros with args: with args 1`] = `

import { ref } from '../../ref.macro';

const foo = ref( 'foo' );
const foo1 = ref( 'foo' );

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = 0;
const foo1 = 0;

`;

exports[`macros without args: without args 1`] = `

import { ref } from '../../ref.macro';

const foo = ref();
const bar = ref();

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = 0;
const bar = 1;

`;
